### PR TITLE
Maven caching

### PR DESCRIPTION
**Pull request by mahesh-hegde**
This introduces some basic caching of maven dependency download tasks. While maven local caches the dependencies, invoking maven itself can slow down `jnigen` significantly.

This PR implements a simple caching mechanism based on file `last modified` timestamps. We write a JSON file containing `maven_downloads` spec. If any file in target dir has changed, or the spec itself has changed, we reinvoke maven. Otherwise maven command is not invoked.

The cache record can be force-invalidated by passing `--invalidate-caches` on command line.

__Also, when we are at it: should we change default maven dependency download folder to somewhere in `.dart_tool`?__

closes: #92

### CODE REVIEW

File: jnigen/lib/src/tools/maven_tools.dart [hunk 1]
```diff
@@ -121,15 +255,19 @@ ${depDecls.join("\n")}
...
     }
     return MavenDependency(components[0], components[1], components[2]);
   }
+
+  @override
+  String toString() {
+    return '$groupID:$artifactID:$version';
+  }
+
   String groupID, artifactID, version;
```
[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply jnigen/lib/src/tools/maven_tools.dart:272]: Maybe we can add https://dart-lang.github.io/linter/lints/prefer_final_fields.html to our lints in general.

**Comment by HosseinYousefi** [comment]
> Also, when we are at it: should we change default maven dependency download folder to somewhere in .dart_tool?

I think it's not necessary to jam everything in there.

File: jnigen/lib/src/tools/maven_tools.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jnigen/lib/src/tools/maven_tools.dart:272]: Isn't that for private fields?

**Comment by mahesh-hegde** [comment]
Higher level design question: should the list of files be written to the record, just in case we merge a change which changes layout of these directories?

**Comment by HosseinYousefi** [comment]
> Edit: I think it's better to add a cacheVersion private field, so that when we change these things, we can ensure invalidation happens.

Good idea.

Related: Something like a `dart run jnigen:clean` would be nice which cleans all the cache/generated files.

**Comment by mahesh-hegde** [comment]
>Related: Something like a dart run jnigen:clean would be nice which cleans all the cache/generated files.

Should this take a config or clean all folders with a cache record file?

**Comment by HosseinYousefi** [comment]
I think in this specific case getting a config is not necessary.

**Comment by mahesh-hegde** [comment]
> I think in this specific case getting a config is not necessary.

User can change the destination to where downloaded files are stored. (We actually use this functionality in integration test with jackson-core library.)

**Comment by HosseinYousefi** [comment]
In what case would you **not** want to remove everything with `jnigen:clean` and instead remove only the files specified in your config?

**Comment by mahesh-hegde** [comment]
> In what case would you not want to remove everything with jnigen:clean and instead remove only the files specified in your config?

Right. Can this be named `jnigen:clean-caches`?

**Comment by HosseinYousefi** [comment]
> Right. Can this be named `jnigen:clean-caches`?

Sure, it's a bit more verbose but a bit more clear. Why not go with `jnigen:clean`, because it might imply that it also removes the generated bindings?

**Comment by HosseinYousefi** [comment]
That's another issue though, I think we can create another issue and discuss it further there. https://github.com/dart-lang/jnigen/issues/209

**Comment by HosseinYousefi** [comment]
One of the tests randomly gets timed out. You can increase it's time limit to 1.5 or 2x for it not to fail again.

**Comment by mahesh-hegde** [comment]
This PR seems to be causing some problems. Let's hold this for a while.

**Comment by HosseinYousefi** [comment]
>  Let's hold this for a while.

Is this ready now?

**Comment by mahesh-hegde** [comment]
The CI error was a stub build.gradle accidentally being committed.

But I also suspect this can be less stable than expected, because cache invalidation is a classic hard topic. Merging it or not is your choice I guess.

**Comment by HosseinYousefi** [comment]
Closing this for now.

**Pull request closed without merging**